ocamllib/src/agent.ml

```diff
@@ -0,0 +1,854 @@
+(* Functions for agent to apply actions to AST *)
+exception InvalidAction of int
+
+(*
+     Given an zippered AST, apply the action
+     Input:
+       - e : an AST with cursor
+       - action : action applied to e
+     Output:
+       the modified AST
+*)
+let perform_action (tree : Expr.z_t) (action : Action.t) : Expr.z_t =
+  (* Handles actions on type subtrees *)
+  let rec act_on_type (type_tree : Type.z_t) : Type.z_t =
+    let build_type (subtr : Type.t) (shape : Action.shape) : Type.z_t =
+      (* actually creates the desired type at the cursor.*)
+      (*Also handles child wrapping logics *)
+      Cursor
+        (match shape with
+        | TypInt -> Type.TInt
+        | TypBool -> Type.TBool
+        | TypHole -> Type.THole
+        | TypArrow_L -> Type.TArrow (subtr, Type.THole)
+        | TypArrow_R -> Type.TArrow (Type.THole, subtr)
+        | TypList -> Type.TList subtr
+        | TypProd_L -> Type.TProd (subtr, THole)
+        | TypProd_R -> Type.TProd (THole, subtr)
+        | _ ->
+            subtr (* all other shapes are for exprssions which are not valid*))
```

Comment:
Should we have a special type for expressions which are not valid? Maybe an example of how this works would be helpful.